src/cipher/pelz_aes_gcm.c

```diff
@@ -0,0 +1,395 @@
+/**
+ * @file  pelz_aes_gcm.c
+ *
+ * @brief Implements AES GCM for pelz.
+ */
+
+#include "cipher/pelz_aes_gcm.h"
+
+#include <openssl/evp.h>
+#include <openssl/rand.h>
+
+#include "memory_util.h"
+
+//############################################################################
+// aes_gcm_encrypt()
+//############################################################################
+int pelz_aes_gcm_encrypt(unsigned char* key,
+			 size_t key_len,
+			 unsigned char* plain,
+			 size_t plain_len,
+			 unsigned char** iv,
+			 size_t *iv_len,
+			 unsigned char** cipher,
+			 size_t* cipher_len,
+			 unsigned char** tag,
+			 size_t *tag_len)
+{
+
+  // validate non-NULL and non-empty encryption key specified
+  if (key == NULL || key_len == 0)
+  {
+    return 1;
+  }
+
+  // validate non-NULL input plaintext buffer specified
+  if (plain == NULL || plain_len == 0)
+  {
+    return 1;
+  }
+
+  *cipher_len = plain_len;
+  *iv_len     = PELZ_GCM_IV_LEN;
+  *tag_len    = PELZ_GCM_TAG_LEN;
+  
+  *cipher = malloc(*cipher_len);
+  if (*cipher == NULL)
+  {
+    return 1;
+  }
+
+  *iv = malloc(*iv_len);
+  if(*iv == NULL)
+  {
+    free(*cipher);
+    *cipher = NULL;
+    *cipher_len = 0;
+    *iv_len = 0;
+    *tag_len = 0;
+    return 1;
+  }
+
+  *tag = (unsigned char*)malloc(*tag_len);
+  if(*tag == NULL)
+  {
+    free(*cipher);
+    free(*iv);
+    *cipher = NULL;
+    *iv = NULL;
+    *cipher_len = 0;
+    *iv_len = 0;
+    *tag_len = 0;
+    return 1;
+  }
+
+  // variable to hold length of resulting CT - OpenSSL insists this be an int
+  int ciphertext_len = 0;
+
+  // initialize the cipher context to match cipher suite being used
+  EVP_CIPHER_CTX *ctx;
+
+  if (!(ctx = EVP_CIPHER_CTX_new()))
```

Comment:
It seems odd to me to do all your mallocs before running EVP_CIPHER_CTX_new and EVP_EncryptInit, which can fail before any of your pointers can be used. Given how many functions error out, it might make further maintenance easier if you allocate/deallocate as you need them.

Reply:
I moved the allocations to right before they're needed in 515f41e.

Reply:
In 5829857 I moved setting default values for the output variables to the top of the functions.